src/main/java/net/neoforged/neoforge/common/crafting/ChildBasedIngredient.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import it.unimi.dsi.fastutil.ints.IntComparators;
+import it.unimi.dsi.fastutil.ints.IntList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import org.jetbrains.annotations.Nullable;
+
+/** Intermediary class for easing handling of ingredients that make use of multiple children */
+public abstract class ChildBasedIngredient extends Ingredient {
+    protected final List<Ingredient> children;
+    private final boolean isSimple;
+    private final boolean synchronizeWithContents;
+
+    @Nullable
+    private ItemStack[] filteredMatchingStacks;
+    @Nullable
+    private IntList packedMatchingStacks;
+
+    protected ChildBasedIngredient(Stream<? extends Ingredient.Value> values, Supplier<? extends IngredientType<?>> type, List<Ingredient> children) {
+        super(values, type);
+        this.children = Collections.unmodifiableList(children);
+        this.isSimple = children.stream().allMatch(Ingredient::isSimple);
+        this.synchronizeWithContents = children.stream().anyMatch(Ingredient::synchronizeWithContents);
+    }
+
+    protected abstract Stream<ItemStack> generateMatchingStacks();
+
+    protected abstract boolean testNonSynchronized(@Nullable ItemStack stack);
+
+    protected abstract IntList generateStackingIds();
+
+    @Override
+    public final ItemStack[] getItems() {
+        if (synchronizeWithContents())
+            return super.getItems();
+
+        if (this.filteredMatchingStacks == null) {
+            this.filteredMatchingStacks = generateMatchingStacks()
+                    .distinct()//Mimic super that calls distinct on the stacks
+                    .toArray(ItemStack[]::new);
+        }
+        return this.filteredMatchingStacks;
+    }
+
+    @Override
+    public final boolean test(@Nullable ItemStack stack) {
+        return synchronizeWithContents() ? super.test(stack) : testNonSynchronized(stack);
+    }
+
+    @Override
+    public final IntList getStackingIds() {
```

Comment:
It shouldn't be necessary to override this method since vanilla computes and caches this information based on `getItems()` already.

Reply:
I know at least historically at one point some mods had implementations of custom ingredients that couldn't represent all possible valid types in `getItems`, granted I am not sure if there are any cases where stacking ids could be done but there were no representations for `getItems`. Personally I don't think it is ideal, but I am unsure if there are any mods that have relied on this aspect of custom ingredients.

Reply:
If you can't represent all possible Items, then you can't possibly represent the list of item IDs either? 🤔

`getStackingIds` should only be used for simple ingredients, and for those there is a trivial conversion between the item IDs and the item stacks.